webapp/src/app.tsx

```diff
@@ -1385,56 +1385,53 @@ export class ProjectView
             pxt.log(`multi-tab sync before load`)
-            p = p.then(() => workspace.syncAsync().then(() => { }))
+            await workspace.syncAsync();
         }
 
         // Try a quick cloud fetch. If it doesn't complete within X second(s),
         // continue on.
+        const TIMEOUT_MS = 15000;
         const timeoutStart = Util.nowSeconds();
         let timedOut = false;
-        p = p.then(() => {
-            Promise.race([
-                pxt.U.delay(1500)
-                    .then(() => { timedOut = true }),
-                cloud.syncAsync([h])
-                    .then(changes => {
-                        if (changes.length) {
-                            const elapsed = Util.nowSeconds() - timeoutStart;
-                            if (timedOut) {
-                                // We are too late; the editor has already been loaded.
-                                // Call the onChanges handler to update the editor.
-                                pxt.tickEvent(`identity.syncOnProjectOpen.timedout`, { 'elapsedSec': elapsed })
-                                if (changes.some(header => header.id === h.id))
-                                    cloud.forceReloadForCloudSync()
-                            } else {
-                                // We're not too late, update the local var so that the
-                                // first load has the new info.
-                                pxt.tickEvent(`identity.syncOnProjectOpen.syncSuccess`, { 'elapsedSec': elapsed })
-                                h = workspace.getHeader(h.id)
-                            }
+        await Promise.race([
+            pxt.U.delay(TIMEOUT_MS)
+                .then(() => { timedOut = true }),
+            cloud.syncAsync([h])
+                .then(changes => {
+                    if (changes.length) {
+                        const elapsed = Util.nowSeconds() - timeoutStart;
+                        if (timedOut) {
+                            // We are too late; the editor has already been loaded.
+                            // Call the onChanges handler to update the editor.
+                            pxt.tickEvent(`identity.syncOnProjectOpen.timedout`, { 'elapsedSec': elapsed })
+                            if (changes.some(header => header.id === h.id))
+                                cloud.forceReloadForCloudSync()
+                        } else {
+                            // We're not too late, update the local var so that the
+                            // first load has the new info.
+                            pxt.tickEvent(`identity.syncOnProjectOpen.syncSuccess`, { 'elapsedSec': elapsed })
+                            h = workspace.getHeader(h.id)
                         }
-                    })
-            ]);
-        });
+                    }
+                })
+        ]);
 
-        return p.then(() => {
+        if (h) {
             workspace.acquireHeaderSession(h);
```

Comment:
@riknoll This code was executing before the `then` containing `Promise.race` above. There must be a reasonable explanation.

Reply:
hmmm yeah the code looks fine to me. no idea what it could be

Reply:
unless it's getting called more than once and interleaving